### ISSUE TITLE
Trying several IP addresses are not implemented

### ISSUE THREAD

**Issue by szepeviktor** [phar-io/phive#195]
See src/services/key/gpg/GnupgKeyDownloader.php, $ipAddress is used only for output.

**Comment by szepeviktor** [comment]
Reopens #184

**Comment by szepeviktor** [comment]
A very dirty solution

```diff
diff --git a/conf/pgp-keyservers.php b/conf/pgp-keyservers.php
index fea20b4..4139dcd 100644
--- a/conf/pgp-keyservers.php
+++ b/conf/pgp-keyservers.php
@@ -1,7 +1,7 @@
 <?php

 return [
+    'hkps.pool.sks-keyservers.net',
     'keys.fedoraproject.org',
     'keyserver.ubuntu.com',
-    'hkps.pool.sks-keyservers.net'
 ];
diff --git a/src/services/key/gpg/GnupgKeyDownloader.php b/src/services/key/gpg/GnupgKeyDownloader.php
index dbc8860..6f44f4c 100644
--- a/src/services/key/gpg/GnupgKeyDownloader.php
+++ b/src/services/key/gpg/GnupgKeyDownloader.php
@@ -42,8 +42,8 @@ class GnupgKeyDownloader implements KeyDownloader {
                 $this->output->writeInfo(\sprintf('Trying %s (%s)', $keyServerName, $ipAddress));

                 try {
-                    $keyInfo   = $this->httpClient->get((new Url('https://' . $keyServerName . self::PATH))->withParams($infoParams));
-                    $publicKey = $this->httpClient->get((new Url('https://' . $keyServerName . self::PATH))->withParams($publicParams));
+                    $keyInfo   = $this->httpClient->get((new Url('https://' . $keyServerName . self::PATH))->withParams($infoParams), null, $ipAddress);
+                    $publicKey = $this->httpClient->get((new Url('https://' . $keyServerName . self::PATH))->withParams($publicParams), null, $ipAddress);

                     $this->ensureSuccess($keyInfo);
                     $this->ensureSuccess($publicKey);
diff --git a/src/shared/http/Curl.php b/src/shared/http/Curl.php
index 9ffc275..ff7f876 100644
--- a/src/shared/http/Curl.php
+++ b/src/shared/http/Curl.php
@@ -13,7 +13,7 @@ class Curl {
     }

     public function setResolve(string $resolveString): void {
-        \curl_setopt($this->curlHandle, \CURLOPT_RESOLVE, $resolveString);
+        \curl_setopt($this->curlHandle, \CURLOPT_RESOLVE, array($resolveString));
     }

     public function addHttpHeaders(array $headers): void {
diff --git a/src/shared/http/CurlHttpClient.php b/src/shared/http/CurlHttpClient.php
index d665f82..5e3afcc 100644
--- a/src/shared/http/CurlHttpClient.php
+++ b/src/shared/http/CurlHttpClient.php
@@ -15,6 +15,9 @@ class CurlHttpClient implements HttpClient {
     /** @var ETag */
     private $etag;

+    /** @var string */
+    private $ip;
+
     /** @var array */
     private $rateLimitHeaders = [];

@@ -48,9 +51,10 @@ class CurlHttpClient implements HttpClient {
     /**
      * @throws HttpException
      */
-    public function get(Url $url, ETag $etag = null): HttpResponse {
+    public function get(Url $url, ETag $etag = null, string $ip = null): HttpResponse {
         $this->url  = $url;
         $this->etag = $etag;
+        $this->ip   = $ip;

         $this->progressHandler->start($url);
         $this->setupCurlInstance();
@@ -92,6 +96,9 @@ class CurlHttpClient implements HttpClient {
         $this->curl->setOptArray($this->config->asCurlOptArray());
         $this->curl->enableProgressMeter([$this, 'handleProgressInfo']);
         $this->curl->setHeaderFunction([$this, 'handleHeaderInput']);
+        if ($this->ip !== null) {
+            $this->curl->setResolve($this->url->getHostname() . ':443:' . $this->ip);
+        }

         $headers = [];

diff --git a/src/shared/http/HttpClient.php b/src/shared/http/HttpClient.php
index 768d96e..d093cc8 100644
--- a/src/shared/http/HttpClient.php
+++ b/src/shared/http/HttpClient.php
@@ -2,7 +2,7 @@
 namespace PharIo\Phive;

 interface HttpClient {
-    public function get(Url $url, ETag $etag = null): HttpResponse;
+    public function get(Url $url, ETag $etag = null, string $ip = null): HttpResponse;

     public function head(Url $url, ETag $etag = null): HttpResponse;
 }
diff --git a/src/shared/http/RetryingHttpClient.php b/src/shared/http/RetryingHttpClient.php
index 9f42bd3..1a3ddfa 100644
--- a/src/shared/http/RetryingHttpClient.php
+++ b/src/shared/http/RetryingHttpClient.php
@@ -40,17 +40,17 @@ class RetryingHttpClient implements HttpClient {
         return $this->doTry('head', $url, $etag);
     }

-    public function get(Url $url, ETag $etag = null): HttpResponse {
+    public function get(Url $url, ETag $etag = null, string $ip = null): HttpResponse {
         $this->triesPerformed = 0;

-        return $this->doTry('get', $url, $etag);
+        return $this->doTry('get', $url, $etag, $ip);
     }

-    private function doTry($method, Url $url, ETag $etag = null) {
+    private function doTry($method, Url $url, ETag $etag = null, string $ip = null) {
         try {
             $this->triesPerformed++;

-            return $this->client->$method($url, $etag);
+            return $this->client->$method($url, $etag, $ip);
         } catch (HttpException $e) {
             if ($this->triesPerformed < $this->maxTries && isset(self::retryCodes[$e->getCode()])) {
                 $this->output->writeInfo(
@@ -58,7 +58,7 @@ class RetryingHttpClient implements HttpClient {
                 );
                 \sleep(2);

-                return $this->doTry($method, $url, $etag);
+                return $this->doTry($method, $url, $etag, $ip);
             }

             throw $e;
```

**Comment by theseer** [comment]
Wow. How embarrassingly stupid. How could I have added `setResolve` and then not even actually make use of it?

While the solution obviously works, I'm not a fan of optional parameters.
I think I'd prefer having the resolved IP become part of a special URL variant.

I'll think about it for a sec and see how that would look.

**Comment by szepeviktor** [comment]
> the resolved IP become part of a special URL variant.

👍

**Comment by theseer** [comment]
Thanks for looking into this btw :)

**Comment by szepeviktor** [comment]
You're welcome.
Please reconsider using `phpstan analise -l max`

**Comment by szepeviktor** [comment]
You've got two 👀 I've got two 👀 PHPStan has 1000 👀 that is 1002 eyes.

**Comment by theseer** [comment]
An alternative approach has been implemented with commit fda19494898977600a1d0045af7c9bc683d21606. 

We'll see what needs tweaking here or if it could be simplified.
